sklearn/partial_dependence.py

```diff
@@ -0,0 +1,571 @@
+    # - 1 for non-multioutput regression and binary classification (shape is
+    #   already correct in those cases)
+    # - n_tasks for multi-output regression
+    # - n_classes for multiclass classification.
+    averaged_predictions = np.array(averaged_predictions).T
+    if is_regressor(est) and averaged_predictions.ndim == 1:
+        # non-multioutput regression, shape is (n_points,)
+        averaged_predictions = averaged_predictions.reshape(1, -1)
+    elif is_classifier(est) and averaged_predictions.shape[0] == 2:
+        # Binary classification, shape is (2, n_points).
+        # we output the effect of **positive** class
+        averaged_predictions = averaged_predictions[1]
+        averaged_predictions = averaged_predictions.reshape(1, -1)
+
+    return averaged_predictions
+
+
+def partial_dependence(est, target_variables, grid=None, X=None,
+                       percentiles=(0.05, 0.95), grid_resolution=100,
+                       method='auto'):
+    """Partial dependence of ``target_variables``.
+
+    Read more in the :ref:`User Guide <partial_dependence>`.
+
+    Parameters
+    ----------
+    est : BaseEstimator
+        A fitted classification or regression model. Multioutput-multiclass
+        classifiers are not supported.
+    target_variables : list or array-like of int
+        The target features for which the partial dependency should be
+        computed.
+    grid : array-like, shape=(n_points, len(target_variables)), optional
+        The grid of values for which the partial dependence should be
+        evaluated. If `None`, the grid will be generated from the values in
+        ``X``.
+    X : array-like, shape=(n_samples, n_features)
+        ``X`` is used both to generate a grid for the ``target_variables``
+        (if ``grid`` is None), and to compute the averaged predictions for
+        the 'brute' method. Optional if ``grid`` is not None and ``method``
+        is 'recursion'.
+    percentiles : tuple of float, optional (default=(0.05, 0.95))
+        The lower and upper percentile used to create the extreme values
+        for the ``grid``. Only used if ``grid`` is None.
+    grid_resolution : int, optional (default=100)
+        The number of equally spaced points on the grid, for each target
+        feature. Only used if ``grid`` is None.
+    method : str, optional (default='auto')
+        The method used to calculate the averaged predictions:
+
+        - 'recursion' is only supported for objects inheriting from
+          `BaseGradientBoosting`, but is optimal in terms of speed. With
+          this method, ``X`` is optional and is only used to build the grid.
+
+        - 'brute' is supported for any estimator, but is more
+          computationally intensive.
+
+        - If 'auto', then 'recursion' will be used for
+          ``BaseGradientBoosting`` estimators, and 'brute' used for other
+          estimators.
+
+        Unlike the 'brute' method, 'recursion' does not account for the
```

Comment:
Don't we want to fix that?

Reply:
Ideally yes but I don't think this is actually possible: the 'recursion' method does not need (and doesn't use) any data `X` to compute the averaged predictions (partial dependence). It uses the trainset that is implicitly stored in each tree predictor, but there's no way for us to call `gbdt._init_decision_function(.)` (see [here](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/ensemble/gradient_boosting.py#L1583)) because we don't have this `X`.

I think the only way to do it correctly would be to convert the `init` estimator into a tree but that's a whole new story.

A hacky way consists in calling `gbdt._init_decision_function(X)` where `X` is the data passed to `[plot_]partial_dependence` but:
- this `X` is not necessarily the training data
- `X` is optional for `partial_dependence`


Note that this is not a bug, the original author(s) knew about this and this is briefly mentioned in the [User Guide](https://scikit-learn.org/stable/modules/ensemble.html#id24).

Reply:
right now all the init predictors don't need X, right? you can pass a single row of zeros with the right feature length.
Of course if any init estimator would do non-constant predictions, this would fail.
We could detect if it's a constant init estimator and if not we would need users to pass the data, which conceptually makes sense.
Actually, if it's not constant we can't use the recursive method any more, right? So in that case we should throw and error anyway. We could use the brute method on the init estimator and the recursive on the trees in that case but not sure if it's even worth it instead of using brute all the way.

Reply:
sorry, it's actually not even using the number of features, just the number of samples to predict on... which we can set to 1.

Reply:
I have added warnings in the code and in the doc.